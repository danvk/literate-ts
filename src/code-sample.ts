import _ from 'lodash';

import {CodeSample, PrefixedCodeSample, Prefix, IdMetadata} from './types';
import {log} from './logger';
import {fail} from './test-tracker';
import {extractAsciidocSamples} from './asciidoc';
import {extractMarkdownSamples} from './markdown';
import {generateIdMetadata} from './metadata';

export interface Processor {
  setLineNum(line: number): void;
  setHeader(header: string): void;
  setDirective(directive: string): void;
  setNextId(idMetadata: IdMetadata): void;
  setNextLanguage(lang: string | null): void;
  addSample(code: string): void;
  resetWithNormalLine(): void;
}

function process(
  text: string,
  slug: string,
  sourceFile: string,
  processor: (filename: string, text: string, processor: Processor) => void,
): PrefixedCodeSample[] {
  const samples: PrefixedCodeSample[] = [];

  let lineNum = 0;
  let lastSectionHeader: string | null = null;
  let lastMetadata: IdMetadata | null = null;
  let lastLanguage: string | null = null;
  let prefixes: readonly Prefix[] = [];
  let skipNext = false;
  let prependNext = false;
  let prependLines: number[] | null = null;
  let nodeModules: readonly string[] = [];
  let tsOptions: {[key: string]: string | boolean} = {};
  let nextIsTSX = false;
  let nextShouldCheckJs = false;

  const p: Processor = {
    setLineNum(line) {
      lineNum = line;
    },
    setHeader(header) {
      p.setDirective('reset');
      lastSectionHeader = header;
      lastMetadata = null;
    },
    setDirective(directive) {
      if (directive === 'reset') {
        prefixes = [];
        prependNext = false;
        skipNext = false;
        tsOptions = {};
        nodeModules = [];
        nextIsTSX = false;
        nextShouldCheckJs = false;
      } else if (directive === 'prepend-to-following') {
        prependNext = true;
      } else if (directive.startsWith('prepend-subset-to-following:')) {
        prependNext = true;
        prependLines = directive
          .split(':', 2)[1]
          .split('-')
          .map(Number);
      } else if (directive.startsWith('prepend-id-to-following')) {
        prefixes = prefixes.concat([
          {
            id: directive
              .split(':')
              .slice(1)
              .join(':'),
          },
        ]);
      } else if (directive.startsWith('skip')) {
        skipNext = true;
      } else if (directive.startsWith('tsconfig:')) {
        const [key, value] = directive.split(':', 2)[1].split('=', 2);
        tsOptions[key] = value === 'true' ? true : value === 'false' ? false : value;
      } else if (directive.startsWith('include-node-module:')) {
        const value = directive.split(':', 2)[1];
        nodeModules = nodeModules.concat([value]);
      } else if (directive === 'next-is-tsx') {
        nextIsTSX = true;
      } else if (directive === 'check-js') {
        nextShouldCheckJs = true;
        tsOptions['allowJs'] = true; // convenience, it's useless without this!
        tsOptions['noEmit'] = true;
      } else {
        throw new Error(`Unknown directive: ${directive}`);
      }
    },
    setNextId(id) {
      lastMetadata = id;
    },
    setNextLanguage(lang) {
      lastLanguage = lang;
    },
    addSample(content) {
      if (
        !lastMetadata &&
        (lastLanguage === 'ts' || (lastLanguage === 'js' && nextShouldCheckJs))
      ) {
        // TS samples get checked even without IDs.
        lastMetadata = generateIdMetadata(slug + '-' + lineNum, sourceFile, lineNum);
      }
      if (lastMetadata) {
        if (!skipNext) {
          samples.push({
            ...lastMetadata,
            sectionHeader: lastSectionHeader,
            language: lastLanguage as CodeSample['language'],
            content,
            prefixes,
            nodeModules,
            isTSX: nextIsTSX,
            checkJS: nextShouldCheckJs,
            sourceFile,
            tsOptions: {...tsOptions},
          });
        }
        if (prependNext) {
          prefixes = prefixes.concat([
            {
              id: lastMetadata.id,
              ...(prependLines ? {lines: prependLines} : {}),
            },
          ]);
          prependNext = false;
          prependLines = null;
        }
      }
    },
    resetWithNormalLine() {
      lastMetadata = null;
      lastLanguage = null;
      skipNext = false;
      nextIsTSX = false;
      nextShouldCheckJs = false;
    },
  };

  processor(sourceFile, text, p);

  return samples;
}

/** Apply HIDE..END and COMPRESS..END directives */
export function stripSource(source: string) {
  // g = match all
  // m = make ^ match at the start of lines (not just start of string)
  // s = let . match newlines
  return source
    .replace(/^ *\/\/ HIDE.*?^ *\/\/ END\n?/gms, '')
    .replace(/^( *\/\/) COMPRESS.*?^\1 END/gms, '$1 ...')
    .replace(/\/\* COMPRESS .* END \*\//g, '/* ... */');
}

/**
 * Verify that the sample in the book matches the alternative in the file.
 *
 * The idea is that the version in the file is an expanded version of what's in the book,
 * something that can actually be run.
 */
export function checkSource(sample: CodeSample, source: string) {
  // Strip out code behind HIDE..END markers
  const strippedSource = stripSource(source);
  if (sample.content.trim() !== strippedSource.trim()) {
    fail('Inline sample does not match sample in source file', sample);
    log('Inline sample:');
    log(sample.content.trim());
    log('----');
    log('Stripped source file sample:');
    log(strippedSource.trim() + '\n');
    return false;
  }
  return true;
}

export function applyPrefixes(
  samples: PrefixedCodeSample[],
  sources: {[id: string]: string} = {},
): CodeSample[] {
  const idToSample = _.keyBy(samples, 'id');
  const sliceLines = (text: string, lines: number[] | undefined) =>
    lines
      ? text
          .split('\n')
          .slice(lines[0] - 1, lines[1])
          .join('\n')
      : text;
  return samples.map(sample => {
    const prefixes = sample.id.endsWith('-output') ? [] : sample.prefixes;
    const content = prefixes
      .map(({id, lines}) => sliceLines(sources[id] || idToSample[id].content, lines))
      .concat([sources[sample.id] || sample.content])
      .join('\n');
    return {
      descriptor: sample.descriptor,
      id: sample.id,
      sectionHeader: sample.sectionHeader,
      language: sample.language,
      tsOptions: sample.tsOptions,
      nodeModules: sample.nodeModules,
      isTSX: sample.isTSX,
      checkJS: sample.checkJS,
      sourceFile: sample.sourceFile,
      content,
    };
  });
}

export function extractSamples(text: string, slug: string, sourceFile: string) {
  let processor;
  if (sourceFile.endsWith('.asciidoc')) {
    processor = extractAsciidocSamples;
  } else if (sourceFile.endsWith('.md')) {
    processor = extractMarkdownSamples;
  } else {
    throw new Error(`Unknown source format, expected .{asciidoc,md}: ${sourceFile}`);
  }

  return process(text, slug, sourceFile, processor);
}
