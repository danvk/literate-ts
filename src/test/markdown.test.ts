import fs from 'fs';
import glob from 'glob';
import path from 'path';

import {extractSamples} from '../code-sample.js';

describe('markdown', () => {
  it('should match snapshots', () => {
    const inputFiles = glob.sync('./src/test/inputs/*.md');

    for (const inputFile of inputFiles) {
      const {base, name} = path.parse(inputFile);
      expect(extractSamples(fs.readFileSync(inputFile, 'utf8'), name, base)).toMatchSnapshot(name);
    }
  });
});
