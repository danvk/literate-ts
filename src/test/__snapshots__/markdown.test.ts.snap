// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown should match snapshots: commented-sample 1`] = `
Array [
  Object {
    "checkJS": false,
    "content": "function somethingDangerous() {}",
    "descriptor": "./commented-sample.md:5",
    "id": "commented-sample-5",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [],
    "sectionHeader": null,
    "sourceFile": "commented-sample.md",
    "tsOptions": Object {},
  },
  Object {
    "checkJS": false,
    "content": "let val = null;  // Type is any
try {
  somethingDangerous();
  val = 12;
  val  // Type is number
} catch (e) {
  console.warn('alas!');
}
val  // Type is number | null",
    "descriptor": "./commented-sample.md:10",
    "id": "commented-sample-10",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [
      Object {
        "id": "commented-sample-5",
      },
    ],
    "sectionHeader": null,
    "sourceFile": "commented-sample.md",
    "tsOptions": Object {},
  },
]
`;

exports[`markdown should match snapshots: doc1 1`] = `
Array [
  Object {
    "checkJS": false,
    "content": "function greet(who: string) {
  console.log('Hello', who);
}",
    "descriptor": "./doc1.md:2",
    "id": "greet-ts",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [],
    "sectionHeader": null,
    "sourceFile": "doc1.md",
    "tsOptions": Object {},
  },
  Object {
    "checkJS": false,
    "content": "console.log(\\"end of file\\");
",
    "descriptor": "./doc1.md:19",
    "id": "doc1-19",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [],
    "sectionHeader": null,
    "sourceFile": "doc1.md",
    "tsOptions": Object {},
  },
]
`;

exports[`markdown should match snapshots: multilinetype 1`] = `
Array [
  Object {
    "checkJS": false,
    "content": "const o = {x: 1, y: 2};
// type is {
//   x: number;
//   y: number;
// }",
    "descriptor": "./multilinetype.md:3",
    "id": "multilinetype-3",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [],
    "sectionHeader": null,
    "sourceFile": "multilinetype.md",
    "tsOptions": Object {},
  },
  Object {
    "checkJS": false,
    "content": "function addWithExtras(a: number, b: number) {
  const c = a + b;  // type is number
  // ...
  return c;
}",
    "descriptor": "./multilinetype.md:13",
    "id": "multilinetype-13",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [],
    "sectionHeader": null,
    "sourceFile": "multilinetype.md",
    "tsOptions": Object {},
  },
]
`;

exports[`markdown should match snapshots: noid 1`] = `
Array [
  Object {
    "checkJS": false,
    "content": "console.log('Hello', 'TS');",
    "descriptor": "./noid.md:7",
    "id": "noid-7",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [],
    "sectionHeader": null,
    "sourceFile": "noid.md",
    "tsOptions": Object {},
  },
]
`;

exports[`markdown should match snapshots: prepend 1`] = `
Array [
  Object {
    "checkJS": false,
    "content": "type AB = 'a' | 'b';",
    "descriptor": "./prepend.md:1",
    "id": "prefix",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [],
    "sectionHeader": null,
    "sourceFile": "prepend.md",
    "tsOptions": Object {},
  },
  Object {
    "checkJS": false,
    "content": "const a: AB = 'a';",
    "descriptor": "./prepend.md:6",
    "id": "combined",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [
      Object {
        "id": "prefix",
      },
    ],
    "sectionHeader": null,
    "sourceFile": "prepend.md",
    "tsOptions": Object {},
  },
  Object {
    "checkJS": false,
    "content": "const a: AB = 'a';",
    "descriptor": "./prepend.md:13",
    "id": "final",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [],
    "sectionHeader": null,
    "sourceFile": "prepend.md",
    "tsOptions": Object {},
  },
]
`;

exports[`markdown should match snapshots: prepend-multiple 1`] = `
Array [
  Object {
    "checkJS": false,
    "content": "type AB = 'a' | 'b';",
    "descriptor": "./prepend-multiple.md:2",
    "id": "prepend-multiple-2",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [],
    "sectionHeader": null,
    "sourceFile": "prepend-multiple.md",
    "tsOptions": Object {},
  },
  Object {
    "checkJS": false,
    "content": "type ABC = AB | 'c';",
    "descriptor": "./prepend-multiple.md:7",
    "id": "prepend-multiple-7",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [
      Object {
        "id": "prepend-multiple-2",
      },
    ],
    "sectionHeader": null,
    "sourceFile": "prepend-multiple.md",
    "tsOptions": Object {},
  },
  Object {
    "checkJS": false,
    "content": "const c: ABC = 'c';",
    "descriptor": "./prepend-multiple.md:11",
    "id": "prepend-multiple-11",
    "isTSX": false,
    "language": "ts",
    "nodeModules": Array [],
    "prefixes": Array [
      Object {
        "id": "prepend-multiple-2",
      },
      Object {
        "id": "prepend-multiple-7",
      },
    ],
    "sectionHeader": null,
    "sourceFile": "prepend-multiple.md",
    "tsOptions": Object {},
  },
]
`;

exports[`markdown should match snapshots: skip 1`] = `Array []`;
